vault/ha.go

```diff
@@ -236,31 +237,76 @@ func (c *Core) StepDown(req *logical.Request) (retErr error) {
 	return retErr
 }
 
-// runStandby is a long running routine that is used when an HA backend
-// is enabled. It waits until we are leader and switches this Vault to
-// active.
+// runStandby is a long running process that manages a number of the HA
+// subsystems.
 func (c *Core) runStandby(doneCh, manualStepDownCh, stopCh chan struct{}) {
 	defer close(doneCh)
 	defer close(manualStepDownCh)
 	c.logger.Info("entering standby mode")
 
-	// Monitor for key rotation
-	keyRotateDone := make(chan struct{})
-	keyRotateStop := make(chan struct{})
-	go c.periodicCheckKeyUpgrade(context.Background(), keyRotateDone, keyRotateStop)
-	// Monitor for new leadership
-	checkLeaderDone := make(chan struct{})
-	checkLeaderStop := make(chan struct{})
-	go c.periodicLeaderRefresh(checkLeaderDone, checkLeaderStop)
-	defer func() {
-		c.logger.Debug("closed periodic key rotation checker stop channel")
-		close(keyRotateStop)
-		<-keyRotateDone
-		close(checkLeaderStop)
-		c.logger.Debug("closed periodic leader refresh stop channel")
-		<-checkLeaderDone
-		c.logger.Debug("periodic leader refresh returned")
-	}()
+	var g run.Group
+	{
+		// This will cause all the other actors to close when the stop channel
+		// is closed.
+		g.Add(func() error {
+			<-stopCh
+			return nil
+		}, func(error) {})
+	}
+	{
+		// Monitor for key rotation
+		keyRotateDone := make(chan struct{})
+		keyRotateStop := make(chan struct{})
+
+		g.Add(func() error {
+			c.periodicCheckKeyUpgrade(context.Background(), keyRotateDone, keyRotateStop)
+			return nil
+		}, func(error) {
+			close(keyRotateStop)
+			c.logger.Debug("shutting down periodic key rotation checker")
+			<-keyRotateDone
+		})
+	}
+	{
+		// Monitor for new leadership
+		checkLeaderDone := make(chan struct{})
+		checkLeaderStop := make(chan struct{})
+
+		g.Add(func() error {
+			c.periodicLeaderRefresh(checkLeaderDone, checkLeaderStop)
+			return nil
+		}, func(error) {
+			close(checkLeaderStop)
+			c.logger.Debug("shutting down periodic leader refresh")
+			<-checkLeaderDone
+		})
+	}
+	{
+		// Wait for leadership
+		leaderDoneCh := make(chan struct{})
+		leaderStopCh := make(chan struct{})
+
+		g.Add(func() error {
+			c.waitForLeadership(leaderDoneCh, manualStepDownCh, leaderStopCh)
+			return nil
+		}, func(error) {
+			close(leaderStopCh)
+			c.logger.Debug("shutting down leader elections")
+			<-leaderDoneCh
+		})
+	}
+
+	// Start all the actors
+	g.Run()
```

Comment:
I'd not seen this library! It's pretty cool. 

All the logic above reads really clearly and has a great pattern. I like it. My only question is around this: https://github.com/oklog/run/blob/master/group.go#L26. "When the first actor returns, all others are interrupted." So when someone finishes first, the first func is aborted and the second func is immediately executed? This behavior seems unclear to me and I'm wondering if you happen to have tested that out and if you could explain what that translates to IRL. I would _imagine_ that the execute func would still finish, so that comment in the library confuses me.

Reply:
Sure! So it runs all the execute funcs in go routines, then when one of them finishes it terminates all the other go routines by calling each one's interrupt func. It allows us to make sure the lifecycle of any of these go routines are bound to the lifecycle of all the other ones. So when one errors or completes the rest will be shutdown.

Reply:
Oh duh, that seems obvious now that you say it! :-)